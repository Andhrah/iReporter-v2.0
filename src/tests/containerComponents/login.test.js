import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../store';
import ConnectedLogin, { Login } from '../../components/container/Login';

const props = {
  createIntervention: jest.fn(),
  history: {
    push: jest.fn(n => n),
  }
};

const eventSuccess = {
  target: {
    value: 'alexandra@test.com',
    name: 'email'
  }
};

const eventClearError = {
  target: {
    value: '',
    name: 'email'
  }
};

const eventSubmit = {
  preventDefault: jest.fn(),
};

describe('Login test suite', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <ConnectedLogin {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles validation correctly', () => {
    const wrapper = shallow(
      <Login {...props} />
    );
    expect(wrapper.instance().state.errors.email).toEqual('');
    expect(wrapper.instance().state.errors.password).toEqual('');
    // Email validation
    wrapper.instance().handleValidation({ email: undefined, password: 'password' });
    expect(wrapper.instance().state.errors.email).toEqual('Please enter a valid email');
    // password validation
    wrapper.instance().handleValidation({ email: 'validuser@test.com', password: undefined });
    expect(wrapper.instance().state.errors.password).toEqual('Invalid password, password must not be less than 6 characters');
  });

  it('handles change correctly', () => {
    const wrapper = shallow(
      <Login {...props} />
    );
    // Success
    expect(wrapper.instance().state.values.email).toEqual('');
    wrapper.instance().handleChange(eventSuccess);
    expect(wrapper.instance().state.values.email).toEqual(eventSuccess.target.value);
    
    // Clear Error
    wrapper.setState({
      errors: {
        email: 'Invalid email'
      }
    });
    wrapper.instance().handleChange(eventClearError);
    expect(wrapper.instance().state.errors.email).toEqual(eventClearError.target.value);
  });

  it('handles submmit correctly', async () => {
    const wrapper = shallow(
      <Login {...props} />
    );
    wrapper.setState({
      values: {
        email: 'validuser@test.com',
        password: 'password'
      }
    });
    expect(wrapper.instance().state.signInCompleted).toEqual(false);
    await wrapper.instance().handleSubmit(eventSubmit);
    await expect(wrapper.instance().state.signInCompleted).toEqual(true);
  });
});
