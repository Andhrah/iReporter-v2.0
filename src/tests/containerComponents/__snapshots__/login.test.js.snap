// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login test suite renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MemoryRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "mycnuj",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "mycnuj",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <ConnectFunction
        createIntervention={[MockFunction]}
        history={
          Object {
            "push": [MockFunction],
          }
        }
      >
        <withRouter(Login)
          createIntervention={[MockFunction]}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          login={[Function]}
        >
          <Route>
            <Login
              createIntervention={[MockFunction]}
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "canGo": [Function],
                  "createHref": [Function],
                  "entries": Array [
                    Object {
                      "hash": "",
                      "key": "mycnuj",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    },
                  ],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "index": 0,
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "mycnuj",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "key": "mycnuj",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                }
              }
              login={[Function]}
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
            >
              <Navbar
                email=""
                firstname=""
                lastname=""
                username=""
              >
                <div
                  className="wrapper"
                >
                  <header>
                    <div
                      className="branding"
                    >
                      <Link
                        to="/"
                      >
                        <a
                          href="/"
                          onClick={[Function]}
                        >
                          <img
                            alt="iReporter logo"
                            src={Object {}}
                          />
                        </a>
                      </Link>
                    </div>
                    <nav
                      className="navbar"
                      id="topnav"
                    >
                      <ul>
                        <li>
                          <Link
                            to="/redflags"
                          >
                            <a
                              href="/redflags"
                              onClick={[Function]}
                            >
                              Red-flags
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/interventions"
                          >
                            <a
                              href="/interventions"
                              onClick={[Function]}
                            >
                              Interventions
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/create-redflag"
                          >
                            <a
                              href="/create-redflag"
                              onClick={[Function]}
                            >
                              Report Corruption
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/create-intervention"
                          >
                            <a
                              href="/create-intervention"
                              onClick={[Function]}
                            >
                              Call for Intervention
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/signup"
                          >
                            <a
                              href="/signup"
                              onClick={[Function]}
                            >
                              Sign Up
                              <i
                                className="fa fa-user-plus"
                              />
                            </a>
                          </Link>
                        </li>
                        <li>
                          <Link
                            to="/login"
                          >
                            <a
                              href="/login"
                              onClick={[Function]}
                            >
                              Sign In
                               
                              <i
                                className="fa fa-user"
                              />
                            </a>
                          </Link>
                        </li>
                        <li>
                          <button
                            className="icon"
                            onClick={[Function]}
                            to="javascript:void(0);"
                            type="button"
                          >
                            <i
                              className="fa fa-bars"
                            />
                          </button>
                        </li>
                      </ul>
                    </nav>
                  </header>
                </div>
              </Navbar>
              <Login
                errors={
                  Object {
                    "email": "",
                    "password": "",
                  }
                }
                handleChange={[Function]}
                handleSubmit={[Function]}
                values={
                  Object {
                    "email": "",
                    "password": "",
                  }
                }
              >
                <section
                  className="form-section"
                >
                  <div
                    className="form-container"
                    id="signin-wrapper"
                  >
                    <h1>
                      Sign In
                    </h1>
                    <form
                      onSubmit={[Function]}
                    >
                      <div
                        className="error"
                        id="login-error"
                      />
                      <div
                        className="input-container"
                      >
                        <label
                          htmlFor="email"
                        >
                          Email/Username
                        </label>
                        <input
                          className="text-input"
                          id="email"
                          name="email"
                          onChange={[Function]}
                          required={true}
                          type="email"
                          value=""
                        />
                        <div
                          className="error"
                          id="email-error"
                        />
                      </div>
                      <div
                        className="input-container"
                      >
                        <label
                          htmlFor="password"
                        >
                          Password
                        </label>
                        <input
                          className="text-input"
                          id="password"
                          name="password"
                          onChange={[Function]}
                          required={true}
                          type="password"
                          value=""
                        />
                        <div
                          className="error"
                          id="password-error"
                        />
                      </div>
                      <button
                        className="form-primary-btn"
                        id="login"
                        type="submit"
                      >
                        Sign In
                      </button>
                    </form>
                    <div
                      className="second-btn"
                    >
                      <h4
                        className="form-h4"
                      >
                        Don't have an account?
                      </h4>
                    </div>
                    <p
                      className="repeating-linear"
                    >
                      <Link
                        className="form-secondary-btn"
                        to="signup"
                      >
                        <a
                          className="form-secondary-btn"
                          href="/signup"
                          onClick={[Function]}
                        >
                          Sign Up
                        </a>
                      </Link>
                    </p>
                  </div>
                </section>
              </Login>
            </Login>
          </Route>
        </withRouter(Login)>
      </ConnectFunction>
    </Router>
  </MemoryRouter>
</Provider>
`;
